### ISSUE TITLE
PulseIn not implemented for ESP8266

### ISSUE THREAD

**Issue by jerryneedell** [adafruit/circuitpython#716]
When trying to use a DHT22 on an ESP8266 - this error is returned:
```
>>> import dht22_test
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "dht22_test.py", line 7, in <module>
  File "adafruit_dht.py", line 190, in temperature
  File "adafruit_dht.py", line 143, in measure
  File "adafruit_dht.py", line 106, in _get_pulses
NotImplementedError: 
```

https://github.com/adafruit/Adafruit_CircuitPython_DHT/blob/master/adafruit_dht.py#L106

I did not see an issue for this so I am just creating this to track it.
edited to add link to pulsio code: https://github.com/adafruit/circuitpython/blob/master/ports/esp8266/common-hal/pulseio/PulseIn.c#L37

**Comment by jerryneedell** [comment]
It is also not implemented for the nrf52. Changed title.

**Comment by nickzoic** [comment]
I've started working on this over at https://github.com/nickzoic/micropython/tree/circuitpython-nickzoic-716-pulseio-esp8266 ... still super sketchy at the moment and missing a lot of the API features.

**Comment by tannewt** [comment]
@nickzoic Looks like a good start! You may be interested in my pending changes here too: https://github.com/adafruit/circuitpython/compare/master...tannewt:pulseio_too_fast

It's adding some handling for gracefully failing when the input signal causes the interrupt to swamp the main code.

**Comment by nickzoic** [comment]
Interesting.  I'm not sure if there's an equivalent for the '66 port, but there's possibly less need for it since communications are UART based ... 
I was wondering about the pulse buffer overflow condition: perhaps it'd make more sense to pause if the ring buffer overflows and resume when you pop the ring buffer. At the moment it over-writes the earliest pulses, but quite often you're interested in the start of a packet ... and it'd also solve the swamping problem.

**Comment by nickzoic** [comment]
OK, PulseIn in that branch is looking pretty good now, and can reliably receive IR remote control signals and can be used by the adafruit_dht library to interface to a DHT11.

PulseOut is on ongoing effort mostly because of the terrible state of ESP8266 PWM and microsecond timers.

**Comment by dhalbert** [comment]
split off nrf52 issue into #1043.

**Comment by sommersoft** [comment]
Closing since ESP8266 is fixed by #926, and nRF is split out.

Thanks @nickzoic!
